Add ZooKeeper connection log

antaljanosbenjamin opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add `zookeeper_connection_log` system table to store historical information about ZooKeeper connections.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

nikitamikhaylov commented:
Why not to add a new type to existing `system.zookeeper_log` ?

Algunenano commented:
> Why not to add a new type to existing system.zookeeper_log ?

Zookeeper log is unusable in the long term due to the high volume of messages it stores.

antaljanosbenjamin commented:
Yes, exactly what @Algunenano said. This should be a much more lightweight table that we might even enable in our cloud. Recently both @vdimir run into issues where this would have been a valuable information.

antaljanosbenjamin commented:
Github suggested me to ask Copilot for a review. I cannot stand not to do so. Let's see how good is that :D

antaljanosbenjamin committed:
Add ZooKeeper connection log

antaljanosbenjamin committed:
Simplify zookeeper connection logging

antaljanosbenjamin committed:
Add `reason` column to ZooKeeper connection log

antaljanosbenjamin committed:
Remove unnecessary configs

antaljanosbenjamin committed:
Use logging instead of print

antaljanosbenjamin committed:
Add test

antaljanosbenjamin committed:
Make test repeatable

antaljanosbenjamin committed:
Fix typo

antaljanosbenjamin committed:
Fix logging

antaljanosbenjamin committed:
Make sure even ConfigReloader will reload keeper config in the right order

By changing the normal zookeeper config first we can ensure that the order of events in `zookeeper_connection_log` will match the expected one even in cases when the config reload is triggered automatically while only one of the files are modified.

antaljanosbenjamin reviewed (commented):
One more thing I just realized: we don't log disconnected events in case of server shutdown/restart. However because of the initialization events it should be obvious. That can be added later.

antaljanosbenjamin committed:
Make test simpler by preventing automatic config reloads

antaljanosbenjamin committed:
Avoid `time.sleep`

antaljanosbenjamin committed:
Make index not nullable

antaljanosbenjamin committed:
Add docs

antaljanosbenjamin committed:
Remove extra empty line

Avogar reviewed (approved):
Looks great!

antaljanosbenjamin commented:
I have to fix the test, it failed in private for some reason.

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into zookeeper-connection-log

antaljanosbenjamin committed:
Make `with_replace_config` work well in case of exceptions

antaljanosbenjamin committed:
Formatting

antaljanosbenjamin committed:
Update test with newly enabled `MULTI_WATCHES`

antaljanosbenjamin committed:
Add log entries for already existing ZooKeeper connections after initializing `ZooKeeperConnectionLog`

antaljanosbenjamin commented:
@Avogar I made the test pass also in private with f35e5047b4c19ede0bf5be05d206ac0ad332eedd, so I think the PR should be ready to be merged. Could you please take another look?

antaljanosbenjamin commented:
- `02443_detach_attach_partition` is [flaky](https://github.com/ClickHouse/ClickHouse/issues/54748)
-  `test_async_connect_to_multiple_ips/test.py::test` is also [flaky](https://github.com/ClickHouse/ClickHouse/issues/85431)
- Fuzzer [issue](https://github.com/ClickHouse/ClickHouse/issues/85432)

Making this green manually.
